Release v0.2.0 (#249)
grafana/xk6-browser: release notes/v0.2.0.md

This release adds three new features:
* HTTP Basic Auth support.
* Automatically adjusting viewport insets depending on the operating system.

> * Denying IPs and hostnames.

Reviewer: ```suggestion
* Blocking IPs and hostnames.
```

> Fixes bugs, including a major one in `WaitForFrameNavigation,` and we now handle the extension shutting-down stage more robustly. Improves testability, logging, and user error messages.

Reviewer: I wouldn't want to highlight the `WaitForFrameNavigation` fix; it's only relevant for dynamic sites, and we want them to be using `waitForSelector()` anyway.

```suggestion
In addition, it comes with a few important bug fixes, and we now handle the extension shutting-down stage more robustly. It also improves testability, logging, and user error messages.
```

  
  Note that it won't update the viewport if the browser runs in headless mode.
  
  See the related issues for more details:
  - [#182: 'screen' property ignored/always set to the same values as 'viewport'](https://github.com/grafana/xk6-browser/issues/182)
  - [#183: Set operating system specific insets](https://github.com/grafana/xk6-browser/issues/183)

## Bugs fixed

- Fix contextWithDoneChan possible goroutine leak. ([#238](https://github.com/grafana/xk6-browser/pull/238))

  We short-circuit the browser communication when the connection is closed. However, sometimes, we may do so when the browser closes or the k6 runner abrupts the test run. This fix is for such situations and helps properly short-circuit the communication when it should end.

- Fix flaky ElementHandle click with detached node. ([#219](https://github.com/grafana/xk6-browser/pull/219))

  Clicking on an element the second time created problems if the first one resulted in an exception.

- Fix Browser.onAttachedToTarget nil pointer. ([#217](https://github.com/grafana/xk6-browser/pull/217))

>   There was a race condition in `Connection` and `FrameSession` when the connection resulted in an error. This change removes running browser attachment and detachment concurrently.

Author: Yeah, but think about future contributors and ourselves. They are useful info.

>   There was a race condition in `Connection` and `FrameSession` when the connection resulted in an error. This change removes running browser attachment and detachment concurrently.

Reviewer: Any contributor can just look at the PRs and Git log if they want further details.

This is why it's a bad idea to generate changelogs from `git log`. Changelogs should be written manually and from the perspective of the user, not the developer.

xk6-browser v0.2.0 is here! :tada:

> This is a minor release with a few important features, several bug fixes, stability improvements and overall code cleanup.

Author: ```suggestion
This is a minor release with a few new features, several bug fixes, stability improvements, and overall code cleanup.
```

## New features

- Support for HTTP Basic Authentication. ([#242](https://github.com/grafana/xk6-browser/pull/242))

  The `BrowserContext`'s `httpCredentials` option previously wasn't working ([#203](https://github.com/grafana/xk6-browser/issues/203)).

  The change also safely handles servers that do not adhere to the basic auth protocol by canceling the authorization process on the first failed attempt.

  It also deprecates `BrowserContext.setHttpCredentials`. The correct way of setting the credentials now is creating a new `BrowserContext` with credentials:

  ```javascript
  // Correct
  const context = browser.newContext({
    httpCredentials: {
      username: 'user',
      password: 'pass',
    }
  });

>   // The following is being deprecated

Author: ```suggestion
  // The following is deprecated
```

    username: 'user',
    password: 'pass',
  });
  ```

  If you need to change or clear credentials within the same script, you should create a new `BrowserContext`.

  See the reasoning behind this deprecation [here](https://github.com/microsoft/playwright/issues/2196) and [here](https://github.com/microsoft/playwright/pull/2763).

- Viewport insets now have the correct size on every OS. ([#185](https://github.com/grafana/xk6-browser/pull/185))

  Browser insets change depending on the operating system, and this feature adds support for adjusting insets of the viewport for Windows, Linux, and macOS.

  Note that it won't update the viewport if the browser runs in headless mode.

  See the related issues for more details: [#182](https://github.com/grafana/xk6-browser/issues/182), [#183](https://github.com/grafana/xk6-browser/issues/183).

- Blocking requests based on the remote host's name or IP via k6's [`blacklistIPs`](https://k6.io/docs/using-k6/options/#blacklist-ips) and [`blockHostnames`](https://k6.io/docs/using-k6/options/#block-hostnames) options. ([#204](https://github.com/grafana/xk6-browser/pull/204), [#205](https://github.com/grafana/xk6-browser/pull/205))

>   If the script specifies these options, the browser will block matching requests with an `ERR_BLOCKED_BY_CLIENT` error.

Author: ```suggestion
  The browser will block matching requests with an `ERR_BLOCKED_BY_CLIENT` error if the script specifies these options.
```

- Support for k6's [`hosts` option](https://k6.io/docs/using-k6/options/#hosts) to override DNS resolution with a custom mapping. ([#207](https://github.com/grafana/xk6-browser/pull/207)).

  This has the same effect as passing the [`host-resolver-rules` argument](https://datacadamia.com/web/browser/chrome#dns_resolver) to Chrome via [browser args](https://github.com/grafana/xk6-browser/blob/ae264530b6fbe0285ac9df9b0f4a907b460bb805/examples/browser_args.js).

## Bugs fixed

- `Page.queryAll()` and `Element.queryAll()` not finding elements. ([#195](https://github.com/grafana/xk6-browser/pull/195))

- Possible issue with consecutive clicks on an element if the first click resulted in an exception. ([#219](https://github.com/grafana/xk6-browser/pull/219))

> - We now handle the extension shutting-down stage more robustly, and it shouldn't print any errors. ([#188](https://github.com/grafana/xk6-browser/pull/188))

Author: ```suggestion
- `Page.waitForNavigation` becomes unresponsive if the browser or connection is closed or a user cancels the script. We now handle the extension shutting-down stage more robustly, and it shouldn't become unresponsive. ([#188](https://github.com/grafana/xk6-browser/pull/188))
```

> - We now handle the extension shutting-down stage more robustly, and it shouldn't print any errors. ([#188](https://github.com/grafana/xk6-browser/pull/188), )

Reviewer 2: ```suggestion
- We now handle the extension shutting-down stage more robustly, and it shouldn't print any errors. ([#188](https://github.com/grafana/xk6-browser/pull/188))
```

- Passing arguments to the browser command now works correctly; see [this example](https://github.com/grafana/xk6-browser/blob/ae264530b6fbe0285ac9df9b0f4a907b460bb805/examples/browser_args.js). ([#206](https://github.com/grafana/xk6-browser/pull/206))

- `Page.waitForNavigation()` doesn't timeout anymore for navigation within the same document (e.g. via anchor links or the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API)). ([#247](https://github.com/grafana/xk6-browser/pull/247)).

> - Several internal fixes that improve stability: [#188](https://github.com/grafana/xk6-browser/pull/188), [#217](https://github.com/grafana/xk6-browser/pull/217), [#238](https://github.com/grafana/xk6-browser/pull/238).

Author: We explained #188 before. Should we repeat it here?

```suggestion
- Internal fixes that improve stability: [#217](https://github.com/grafana/xk6-browser/pull/217) and [#238](https://github.com/grafana/xk6-browser/pull/238).
```
